### DISCUSSION TITLE
Asio integration : better futures completion token

### DISCUSSION THREAD

**Discussion by pfeatherstone** [alandefreitas/futures#12] [General]
I can see in https://alandefreitas.github.io/futures/networking/asio/#asio-executors that you provide an asio executor which you can use with your library.

Do you do the opposite? I.e., do  you provide a better futures completion token which can be used in asio such that functions like `asio::async_write()` would return a `cfuture<>` which supports continuations etc. https://github.com/Naios/continuable for example provides an asio completion token which allows you to do that.

On that note, it looks like this library and [continuable](https://github.com/Naios/continuable) have similar goals and do similar things. Have you compared and benchmarked these libraries?

**Comment by alandefreitas** [comment]
Yes. This is very useful. Right now, I'm working on a complete refactor of the operation states that will make them much more reliable/efficient and will also make integration with other libraries much easier.

Writing Asio completion tokens is not really that complex, but I'm planning to integrate Asio both ways only after this refactor. 

> On that note, it looks like this library and [continuable](https://github.com/Naios/continuable) have similar goals and do similar things. Have you compared and benchmarked these libraries?

Yes. I've looked at [continuable](https://github.com/Naios/continuable) and it looks great. I don't have any benchmarks comparing these two libraries for a couple of reasons:

- I already have benchmarks that compare the futures with the executor-level callbacks as a baseline, and there's no benchmark harder than that. The overhead we over on top of the direct callback is already minimal so the difference between libraries cannot be more significant than that.
- Second, my expectation is that the difference between these libraries (once this library is complete) is to be more about the API and features than benchmarks. The direction I have is to develop a system that integrates everything that has been proposed as vocabulary types: [the standard executor model](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0443r13.html), `std::future`-like interfaces with continuations, two-way compatibility with Asio, eager and deferred tasks, heterogenous computing at the executor level (`std::executor_of`) with GPU executors provided by the library, traits to integrate deferred futures with whatever model of sender/receiver is approved in the future, recursive coroutines as tasks for easy representation of task graphs, and standard asynchronous algorithms running on executors instead of just policies.

It goes without saying that completing all of that is going to take some time but that's the direction I'm going :)

Because this refactor of the operation state is quite complex, I haven't been pushing incremental updates to the library. But I'm actually working hard on this one, something I can't say for most of my other libraries at this time.

**Reply by pfeatherstone** [reply]
I look forward to seeing the updates

**Reply by pfeatherstone** [reply]
Don't you reckon heterogeneous programming would be better served with something like the taskflow library ?

**Reply by alandefreitas** [reply]
> Don't you reckon heterogeneous programming would be better served with something like the taskflow library ?

That's also a great library. But it's not the direction I'm going. :)

The idea of heterogeneous programming with the models I mentioned in the previous comment is that the transition would be a little more transparent as long as the executors support the task types you're using.

**Reply by pfeatherstone** [reply]
Just re-read this. So your plan is to basically re-implement https://github.com/NVIDIA/stdexec, and support future-like continuations and asio integration? That's a huge amount of work.

**Reply by alandefreitas** [reply]
I don't have a complete view of what's in stdexec. Maybe the documentation is lacking but as far as I understand stdexec (and libunifex) is an implementation of P2300 with a few extensions to make it more useful or to extend the proposal. I have no intention to reimplement the P2300 functionality here at all. I also have no intention to reimplement any asio functionality. At most I would redefine a few concepts to avoid depending on these libraries.

This library would only provide _optional_ traits needed for integration with P2300 and Asio in case the user needs any of these things in their application. There's no intention to reimplement let alone "compete" with these libraries in any sense. I think the situation is a little like https://github.com/lewissbaker/cppcoro. C++ coroutines define vocabulary concepts for coroutines, and cppcoro defines concrete types within that system to really use the coroutines.  

If you imagine a model where you have executors/callbacks, then futures/promises, then coroutines, one on top of the previous, Asio provides vocabulary concepts for concrete executors/callbacks while P2300 provides vocabulary concepts for concrete lazy futures and promises. This library would provide this concrete functionality just like cppcoro provides the concrete generators we need to use coroutines. The only overlap might be in a few tangential executors or something like that.

**Reply by pfeatherstone** [reply]
Sounds awesome!

**Reply by pfeatherstone** [reply]
@alandefreitas Do you have an update on this? I'm quite keen to try this out.

**Comment by alandefreitas** [comment]
> I'm working on a complete refactor of the operation states

I think this will take a couple of months... after that we'll have the same features it supports now with a much better implementation and much better tests.

My wild guess is that other features will then be added incrementally over a year. Adding features and helpers is surprisingly easy compared to the core implementation.

**Reply by pfeatherstone** [reply]
Do you plan on publishing or presenting a talk ?

**Reply by alandefreitas** [reply]
I have plans to significantly improve the documentation but no talks or papers before a dozen asio and PR2300 experts give me their feedback.

**Reply by pfeatherstone** [reply]
Awesome. I would like to never have to use `std::thread`, `std::mutex`, `std::condition_variable` ever again, build nice task graphs, throw it a thread pool and know for sure that my program is more efficient than anything i could ever write by hand.

**Reply by pfeatherstone** [reply]
I imagine a combination of Asio, this, maybe [Taskflow](https://github.com/taskflow/taskflow), coroutines, are the answer.
